src/lib.rs

```diff
@@ -710,26 +765,56 @@ impl SystemHandle {
     /// Subscribe given `recipient` to events of type `E`. See [`Context::subscribe()`].
     pub fn subscribe_recipient<M: 'static, E: Event + Into<M>>(&self, recipient: Recipient<M>) {
         let mut event_subscribers = self.event_subscribers.write();
+        event_subscribers.subscribe_recipient::<M, E>(recipient);
+    }
 
-        let subs = event_subscribers.events.entry(TypeId::of::<E>()).or_default();
+    /// Subscribe given `recipient` to events of type `E` and send the last cached event to it.
+    /// See [`Context::subscribe_and_receive_latest()`].
+    pub fn subscribe_and_receive_latest<M: 'static, E: CacheableEvent + Into<M>>(
+        &self,
+        recipient: Recipient<M>,
+    ) -> Result<(), SendError> {
+        let mut event_subscribers = self.event_subscribers.write();
 
-        subs.push(Box::new(move |e| {
-            if let Some(event) = e.downcast_ref::<E>() {
-                let msg = event.clone();
-                recipient.send(msg.into())?;
+        // Send the last cached value if there is one.
+        if let Some(last_cached_value) = event_subscribers.last_value_cache.get(&TypeId::of::<E>())
+        {
+            if let Some(msg) = last_cached_value.downcast_ref::<E>() {
+                recipient.send(msg.clone().into())?;
             }
+        }
```

Comment:
I wonder it it would be worth acquiring only a shared access for getting and sending the cached event, and only upgrading to exclusive access for pushing to subscribers.

If we decide to do that we need to be careful not to introduce the possibility of missed events. That depends on how `upgrade()` behaves w.r.t. other writers waiting for exclusive access. It'll work only if `upgrade()` has priority over `write()`. The docs don't seem to mention that so it's likely *not* the case? If not, this could happen

1. `subscribe_and_receive_latest()` acquires a read lock.
2. The cache is empty. No last event to send.
3. `publish()` requests exclusive access. This blocks because `subscribe_and_receive_latest()` has the read lock.
4. `subscribe_and_receive_latest()` tries to `upgrade()`, but `publish()` is already waiting in line and gets exclusive access first.
5. `publish()` saves the event to cache, fans it out to subscribers and releases its lock.
6. `subscribe_and_receive_latest()` is unblocked, finally upgrades and pushes to subscribers. But it's too late.


If we don't do this, we should add a comment explaining why we don't this to prevent future ourselves introducing this subtle bug. Something like:

```
// Require exclusive access for the entire duration of this function to ensure no one
// publishes an event after we checked the cache and before we push to subscribers.
// In such case the event wouldn't be delivered to this subscriber.
```

Reply:
The optimization you describe would be possible with https://docs.rs/parking_lot/latest/parking_lot/type.RwLock.html#method.upgradable_read.

But this is _subscribe_: it is called only once per event and subscriber -> not worth optimizing IMO. (I'll also post a general comment on the PR).